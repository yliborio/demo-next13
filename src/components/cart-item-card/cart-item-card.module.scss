@import "./../../styles/variables.scss";
.container {
  display: flex;
  padding: 1rem;
  border-radius: 8px;
  background-color: $bg-secondary;
  gap: 2rem;
  flex-direction: row;
  @media (max-width: $sm-mobile) {
    flex-direction: column;
  }

  .image {
    display: flex;
    align-items: center;
    justify-content: center;

    > img {
      width: 8vw;
      height: auto;
      @media (max-width: $sm-mobile) {
        width: 200px;
      }
    }
  }
  .info {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    font-size: $font-size-3;

    @media (max-width: $sm-mobile) {
      gap: 1rem;
      max-height: fit-content;
    }

    max-height: 150px;
    .title,
    .price {
      font-size: $font-size-2;
      @media (max-width: $sm-mobile) {
        font-size: $font-size-3;
      }
      font-weight: 600;
    }
    .button {
      background-color: transparent;
      border: none;
      cursor: pointer;
      text-decoration: underline;
      color: rgb(193, 49, 49);
      width: fit-content;
    }

    .quantity {
      display: flex;
      flex-direction: row;
      gap: 5px;
      align-items: center;
      > button {
        cursor: pointer;
        width: 20px;
        height: 20px;
        border: none;
        background-color: $bg-primary;
        border-radius: 100%;
      }
    }
  }
}
