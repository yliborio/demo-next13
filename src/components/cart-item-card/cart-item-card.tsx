import { FakeAPIProduct } from "core/types/product";
import Image from "next/image";

import styles from "./cart-item-card.module.scss";
import { useCart } from "../../hooks/useCart/useCart";

interface CartItemCardProps {
  product: FakeAPIProduct;
  quantity: number;
}

export const CartItemCard = ({ product, quantity }: CartItemCardProps) => {
  const { removeProduct, addProduct, removeQuantity } = useCart();
  const handleChange = () => {
    removeProduct(product);
  };

  return (
    <div className={styles["container"]} data-testid="cart-item-card">
      <div className={styles["image"]}>
        <Image alt="" src={product.image} sizes="100vw" width={0} height={0} />
      </div>
      <div className={styles["info"]}>
        <span className={styles["title"]}>{product.title}</span>
        <span className={styles["price"]}>{`$${product.price}`}</span>
        <button className={styles["button"]} onClick={handleChange}>
          Remove
        </button>
        <div className={styles["quantity"]}>
          <span>Quantity:</span>
          <button onClick={() => removeQuantity(product)}>-</button>
          <span>{`${quantity}`}</span>
          <button onClick={() => addProduct(product)}>+</button>
        </div>
      </div>
    </div>
  );
};
