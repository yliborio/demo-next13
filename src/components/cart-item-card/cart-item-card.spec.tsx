import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { CartItemCard } from "./cart-item-card";
const useCartMock = {
  removeProduct: jest.fn(),
  addProduct: jest.fn(),
  removeQuantity: jest.fn(),
};
jest.mock("../../hooks/useCart/useCart", () => ({
  useCart: () => useCartMock,
}));

describe("CartItemCard Component", () => {
  it("should render a cart item card with correct product and quantity", () => {
    jest.spyOn(React, "useContext").mockImplementation(useCartMock as any);

    const product = {
      id: 1,
      title: "Sample Product",
      price: 10.0,
      image: "/sample-image.jpg",
    } as any;

    const quantity = 2;

    const { asFragment } = render(
      <CartItemCard product={product} quantity={quantity} />
    );

    //match snapshot
    expect(asFragment()).toMatchSnapshot();

    // Assert that the product title and price are rendered
    expect(screen.getByText("Sample Product")).toBeInTheDocument();
    expect(screen.getByText("$10")).toBeInTheDocument();

    // Assert that the remove button is rendered
    const removeButton = screen.getByText("Remove");
    expect(removeButton).toBeInTheDocument();

    fireEvent.click(removeButton);
    expect(useCartMock.removeProduct).toBeCalledWith(product);

    expect(screen.getByText("Quantity:")).toBeInTheDocument();
    expect(screen.getByText("2")).toBeInTheDocument();

    // Assert that the "+" and "-" buttons are rendered
    expect(screen.getByText("+")).toBeInTheDocument();
    expect(screen.getByText("-")).toBeInTheDocument();

    // Simulate a click on the "+" and "-" buttons and assert that the corresponding functions are called
    fireEvent.click(screen.getByText("+"));
    expect(useCartMock.addProduct).toHaveBeenCalledWith(product);

    fireEvent.click(screen.getByText("-"));
    expect(useCartMock.removeQuantity).toHaveBeenCalledWith(product);
  });
});
