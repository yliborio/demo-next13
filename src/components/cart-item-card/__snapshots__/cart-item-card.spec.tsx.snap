// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemCard Component should render a cart item card with correct product and quantity 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="cart-item-card"
  >
    <div
      class="image"
    >
      <img
        alt=""
        data-nimg="1"
        decoding="async"
        height="0"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fsample-image.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Fsample-image.jpg&w=640&q=75 640w, /_next/image?url=%2Fsample-image.jpg&w=750&q=75 750w, /_next/image?url=%2Fsample-image.jpg&w=828&q=75 828w, /_next/image?url=%2Fsample-image.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fsample-image.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fsample-image.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fsample-image.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fsample-image.jpg&w=3840&q=75 3840w"
        style="color: transparent;"
        width="0"
      />
    </div>
    <div
      class="info"
    >
      <span
        class="title"
      >
        Sample Product
      </span>
      <span
        class="price"
      >
        $10
      </span>
      <button
        class="button"
      >
        Remove
      </button>
      <div
        class="quantity"
      >
        <span>
          Quantity:
        </span>
        <button>
          -
        </button>
        <span>
          2
        </span>
        <button>
          +
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
